### PR TITLE
[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section

### PR DESCRIPTION
**Pull request by simpal01**
When the .eh_frame section is placed at a non-zero
offset within its output section, the relocation
value within .eh_frame are computed incorrectly.

We had similar issue in .ARM.exidx section and it has been
fixed already in https://reviews.llvm.org/D148033.

While applying the relocation using S+A-P,  the value
 of P (the location of the relocation) is getting wrong. 
P is:
  P = SecAddr + rel.offset, But SecAddr points to the
starting address of the outputsection rather than the
starting address of the eh frame section within that
output section.

This issue affects all targets which generates .eh_frame 
section. Hence fixing in all the corresponding targets it affecting.

### CODE REVIEW

**Review by smithp35 — COMMENTED** [review]
This will only fix the problem for AArch64, but it also affects other targets. Assuming relocateAlloc is the best place to fix this, then it would need applying in all instances of relocateAlloc.

File: lld/ELF/Arch/AArch64.cpp [hunk 1]
```diff
@@ -770,6 +770,9 @@ void AArch64::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
   uint64_t secAddr = sec.getOutputSection()->addr;
   if (auto *s = dyn_cast<InputSection>(&sec))
     secAddr += s->outSecOff;
+  else if (auto *eh = dyn_cast<EhInputSection>(&sec))
```
[the remark this answers is not in the document]
Comment (simpal01) [reply lld/ELF/Arch/AArch64.cpp:773]: yeah this affects all other targets (might be X86 and PPC as well).

[the remark this answers is not in the document]
Comment (smithp35) [reply lld/ELF/Arch/AArch64.cpp:773]: OK, reading through the code. The .ARM.exidx sections were also InputSections so we could alter their outSecOff to compensate. However EhInputSections are not InputSections so this won't work here. We need to find the outSecOff of the ehframe section (which is usually 0, but with some linker scripts isn't).

I don't know of a legal EhInputSection that would return getParent() == nullptr so it should be possible to remove the nested if.

I suggest use ehIn and ehFrame to make it a bit clearer which sections we are referring to. I've use SyntheticSection as it is what EhInputSection getParent() although that is derived from InputSection so in the code below InputSection would work too.
```
if (auto *ehIn = dyn_cast<EhInputSection>(&sec) {
  SyntheticSection *ehFrame = ehIn->getParent();
  SecAddr += ehFrame->outSecOff; 
}
```

**Review by smithp35 — COMMENTED** [review]
Thanks for fixing the commits, that makes it easier to review.

I can see that this would need a fix in relocateAlloc. I think you'll need to update the generic one in Target.cpp as well as the Target specific ones for x86_64 and PPC (and any other I've missed).

You'll need to add x86_64, PPC and at least one other target's (for the generic Target) so that we've covered all the code-paths. By replacing the nops with a common data directive you might be able to make the same file work for all 3. However I think it would still need separate files so that the REQUIRES only had one target.

File: lld/ELF/SyntheticSections.cpp [hunk 2]
```diff
@@ -583,9 +583,14 @@ static uint64_t readFdeAddr(uint8_t *buf, int size) {
...
   // The starting address to which this FDE applies is
-  // stored at FDE + 8 byte.
+  // stored at FDE + 8 byte. And this offset is within
+  // the .eh_frame section.
   size_t off = fdeOff + 8;
   uint64_t addr = readFdeAddr(buf + off, enc & 0xf);
+  // Adding outSecOff as finalizeAddressDependentContent()
+  // may have altered the corresponding outSecOff. This is
+  // required to get the correct PC relative offset.
+  off = off + outSecOff;
```
Comment (simpal01) [inline lld/ELF/SyntheticSections.cpp:593]: on  a second thought, adding outSecOff inside the condition might be  more sensible.
  if ((enc & 0x70) == DW_EH_PE_pcrel)
    return addr + getParent()->addr + off;

**Review by smithp35 — COMMENTED** [review]
Can you rename the PR and the commit message to remove [AARCH64] as this is now a general change.

I think with this and your own suggestion of moving the change in getFdePc inside the if statement it will be ready for my approval. We'll want to get MaskRay to take a look too.

**Comment by simpal01** [comment]
> Can you rename the PR and the commit message to remove [AARCH64] as this is now a general change.
> 
> I think with this and your own suggestion of moving the change in getFdePc inside the if statement it will be ready for my approval. We'll want to get MaskRay to take a look too.

Done

**Comment by simpal01** [comment]
@MaskRay Could you please have a look into this patch?

File: lld/ELF/Arch/AArch64.cpp [hunk 3]
```diff
@@ -770,6 +770,10 @@ void AArch64::relocateAlloc(InputSectionBase &sec, uint8_t *buf) const {
   uint64_t secAddr = sec.getOutputSection()->addr;
   if (auto *s = dyn_cast<InputSection>(&sec))
     secAddr += s->outSecOff;
+  else if (auto *ehIn = dyn_cast<EhInputSection>(&sec)) {
+    SyntheticSection *ehFrame = ehIn->getParent();
```
Comment (MaskRay) [inline lld/ELF/Arch/AArch64.cpp:774]: `ehFrame` is used only once. Just inline its use and drop the braces. Ditto everywhere.

File: lld/test/ELF/eh-frame-nonzero-offset-aarch64.s [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// REQUIRES: aarch64
+// RUN: rm -rf %t && split-file %s %t
```
Comment (MaskRay) [inline lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:2]: `// RUN: rm -rf %t && split-file %s %t && cd %t` then you can omit `%t/` below.

File: lld/test/ELF/eh-frame-nonzero-offset-aarch64.s [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+// ZERO:      0x00000080 10000000 00000000 017a5200 017c1e01
+// ZERO-NEXT: 0x00000090 1b0c1f00 10000000 18000000 64ffffff
+// ZERO-NEXT: 0x000000a0 08000000 00000000 00000000
+
+//--- eh-frame-non-zero-offset.t
+SECTIONS {
+  .text : { *(.text .text.*) }
+  .eh_frame : {
+  /* Alignment padding within .eh_frame */
+  . = ALIGN(128);
```
Comment (MaskRay) [inline lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:30]: Indent input section descriptions within `.eh_frame`.

`. = ALIGN(128);` may not necessarily add a padding if `.` is already aligned.
Use `. += 128;` instead?

File: lld/test/ELF/eh-frame-nonzero-offset-aarch64.s [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  __eh_frame_end = .;
+  }
+}
+
+//--- eh-frame-zero-offset.t
+SECTIONS {
+  .text : { *(.text .text.*) }
+  .eh_frame : ALIGN(128) {
+  __eh_frame_start = .;
+  *(.eh_frame .eh_frame.*) ;
```
Comment (MaskRay) [inline lld/test/ELF/eh-frame-nonzero-offset-aarch64.s:42]: For the purpose of this test, you can omit `.eh_frame.*`. GNU ld's internal linker script still uses `.eh_frame.*` but it is not clear which port still uses `.eh_frame.*`.

**Comment by MaskRay** [comment]
> [LLD] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.

The convention is to use an imperative sentence and omit the trailing dot.

Perhaps `[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section` will be better.
(Some patches use `[LLD][ELF]` or `[lld][ELF]` but I find `[lld]` redundant.)

**Comment by simpal01** [comment]
> > [LLD] lld incorrectly handles .eh_frame when it has a non-zero offset within its output section.
> 
> The convention is to use an imperative sentence and omit the trailing dot.
> 
> Perhaps `[ELF] Handle relocations in synthetic .eh_frame with a non-zero offset within the output section` will be better. (Some patches use `[LLD][ELF]` or `[lld][ELF]` but I find `[lld]` redundant.)

Done

File: lld/test/ELF/eh-frame-nonzero-offset-aarch64.s (hunk 6 above)
Comment (simpal01) [reply lld/test/ELF/eh-frame-nonzero-offset-aarch64.s]: Done

File: lld/test/ELF/eh-frame-nonzero-offset-aarch64.s (hunk 5 above)
Comment (simpal01) [reply lld/test/ELF/eh-frame-nonzero-offset-aarch64.s]: Yeah looks like that is the case. Replaced **. = ALIGN(128**) with **. += 128**.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**